File: core/src/main/java/io/xj/core/model/segment_meme/SegmentMeme.java
// Copyright (c) 2018, XJ Music Inc. (https://xj.io) All Rights Reserved.
package io.xj.core.model.segment_meme;

import com.google.common.collect.Lists;
import io.xj.core.exception.CoreException;
import io.xj.core.model.meme.Meme;
import io.xj.core.model.segment.Segment;
import io.xj.core.model.segment.SegmentEntity;

import java.math.BigInteger;
import java.util.Collection;
import java.util.Objects;
import java.util.UUID;

/**
 [#166132897] Segment model handles all of its own entities
 */
public class SegmentMeme extends SegmentEntity implements Meme {
  private String name;

  public static SegmentMeme of(String name) {
    return new SegmentMeme().setName(name);
  }

  public static Collection<SegmentMeme> aggregate(Collection<Segment> segments) {
    Collection<SegmentMeme> aggregate = Lists.newArrayList();
    segments.forEach(segment -> aggregate.addAll(segment.getMemes()));
    return aggregate;
  }

  public SegmentMeme setName(String name) {
    this.name = name;
    return this;
  }

  @Override
  public SegmentMeme setUuid(UUID uuid) {
    this.uuid = uuid;
    return this;
  }

  @Override
  public SegmentMeme setSegmentId(BigInteger segmentId) {
    this.segmentId = segmentId;
    return this;
  }

  @Override
  public void validate() throws CoreException {
    super.validate();
    if (Objects.isNull(name) || name.isEmpty()) {
      throw new CoreException("Name is required.");
    }
  }

  public String getName() {
    return name;
  }
}

File: engine/src/entities/segment/SegmentMeme.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include "xjmusic/entities/segment/SegmentMeme.h"

namespace XJ {

  bool SegmentMeme::equals(const SegmentMeme &segmentMeme) const {
    return id == segmentMeme.id &&
           segmentId == segmentMeme.segmentId &&
           name == segmentMeme.name;
  }

  unsigned long long SegmentMeme::hashCode() const {
    return std::hash<std::string>{}(id) ^
           std::hash<int>{}(segmentId) ^
           std::hash<std::string>{}(name);
  }

}// namespace XJ
